src/main/java/com/denys/toukdemo/service/ReservationServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+import com.denys.toukdemo.repository.UserRepository;
+import com.denys.toukdemo.dto.Reservation;
+import com.denys.toukdemo.dto.ReservationStatus;
+import com.denys.toukdemo.dto.ReservationUnit;
+import com.denys.toukdemo.dto.Room;
+import com.denys.toukdemo.mapper.ReservationMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+@Transactional(propagation = Propagation.REQUIRED)
+@Component
+public class ReservationServiceImpl implements ReservationService {
+
+    @Autowired
+    SeansRepository seansRepository;
+
+    @Autowired
+    ReservationRepository reservationRepository;
+
+    @Autowired
+    UserRepository userRepository;
+
+    @Override
+    public Room getRoomWithReservedPlacesBySeansId(String seansId) {
+        SeansEntity seans = seansRepository.findSeansEntityByIdEquals(Integer.parseInt(seansId));
+        Collection<ReservationEntity> reservationEntities = seans.getReservationsById();
+        RoomEntity roomEntity = seans.getRoomByRoomId();
+        Room room = new Room(roomEntity.getRoomType());
+        for (ReservationEntity reservationEntity : reservationEntities) {
+            if(reservationEntity.getStatus() != ReservationStatus.CAMCEL) {
+                Integer row = reservationEntity.getId().getRowNumer();
+                Integer place = reservationEntity.getId().getPlace();
+                room.setPlaceReserve(row, place);
+            }
+        }
+        return room;
+    }
+
+    @Override
+    public void saveReservation(Reservation reservation, String confirmNo) throws RuntimeException {
+        SeansEntity seansEntity = getReservationOrThrowException(reservation);
+        List<ReservationEntity> reservationEntities = createReservationEntityList(reservation, confirmNo);
+        createOrUpdateUsersEntity(reservation, reservationEntities);
+        updateSeansEntity(seansEntity, reservationEntities);
+        reservationRepository.save(reservationEntities);
+    }
+
+    @Override
+    public boolean confirmReservation(String confirmNo) {
+        boolean result = true;
+        List<ReservationEntity> reservedEntities = reservationRepository.findAllById_Id(confirmNo);
+        for (ReservationEntity reservedEntity : reservedEntities) {
+            LocalDateTime reservedTime = reservedEntity.getReservationDttm().toLocalDateTime();
+            if(tooLatePlus15Minutes(reservedTime)) {
+                reservedEntity.setStatus(ReservationStatus.CAMCEL);
+                result = false;
+            }
+            else
+                reservedEntity.setStatus(ReservationStatus.CONFIRM);
+        }
+        reservationRepository.save(reservedEntities);
```

Comment:
jeśli dobrze rozumiem bazę danych - jak tu zapiszesz rezerwacje w statusie CAMCEL i ona będzie miała voucher to on przepada?

Reply:
> jeśli dobrze rozumiem bazę danych - jak tu zapiszesz rezerwacje w statusie CAMCEL i ona będzie miała voucher to on przepada?

tak,  ja zrobiłem takie zalożenia, to raczej pytanie biznesowe którego nie było